src/product/tests/product.service.spec.ts

```diff
@@ -0,0 +1,194 @@
+import {
+  NotFoundException,
+  InternalServerErrorException,
+  BadRequestException,
+} from '@nestjs/common';
+import { Test, TestingModule } from '@nestjs/testing';
+import { getRepositoryToken } from '@nestjs/typeorm';
+import { v4 as uuid } from 'uuid';
+
+import { ProductService } from '../product.service';
+import { Product } from '../entities/product.entity';
+import { CreateProductDto, UpdateProductDto } from '../dto';
+import { ProductState } from '../types/product';
+import { Sorting } from '../../common/types/pagination';
+import { PaginationDto } from '../../common/dto/pagination.dto';
+import { ErrorCodes } from '../../common/types/error';
+
+describe('ProductService', () => {
+  let service: ProductService;
+  let mockRepository;
+
+  beforeEach(async () => {
+    mockRepository = {
+      findOneBy: jest.fn(),
+      find: jest.fn(),
+      create: jest.fn(),
+      save: jest.fn(),
+      preload: jest.fn(),
+      remove: jest.fn(),
+    };
+
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [
+        ProductService,
+        {
+          provide: getRepositoryToken(Product),
+          useValue: mockRepository,
+        },
+      ],
+    }).compile();
+
+    service = module.get<ProductService>(ProductService);
+  });
+
+  it('should be defined', () => {
+    expect(service).toBeDefined();
+  });
+
+  describe('findOne', () => {
+    it('should return a product if it exists', async () => {
+      const product = new Product();
+      mockRepository.findOneBy.mockReturnValue(product);
```

Comment:
WDTY about using the `mockReturnValueOnce`? since it's possible that we forget to mock the function in other tests and get a false positive result

Reply:
good